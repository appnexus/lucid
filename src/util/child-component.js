import React from 'react';
import _ from 'lodash';

/**
 * createChildComponent
 *
 * Returns a React Component which does not render anything, but has some useful static functions
 *
 * @return {ReactComponent} which renders `null` but exposes static helper functions
 */
export function createChildComponent (definition={}) {
	return React.createClass({
		propTypes: _.get(definition, 'propTypes', null),
		displayName: _.get(definition, 'displayName', null),
		render: () => null,
		statics: {

			/**
			 * getOwnChildren
			 *
			 * Returns all collective children of elements with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} All the collective children of elements with this component type
			 */
			getOwnChildren(elements) {
				let component = this;
				let children = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						children = children.concat(React.Children.map(element.props.children, x => x));
					}
				});
				return children;
			},

			/**
			 * getAll
			 *
			 * Returns all elements with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} elements with this component type
			 */
			getAll(elements) {
				let component = this;
				let elementsOfThisType = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						elementsOfThisType = elementsOfThisType.concat(element);
					}
				});
				return elementsOfThisType;
			},

			/**
			 * getAllProps
			 *
			 * Returns the `props` object of each element with this component type
			 *
			 * @param {Array} elements - usually `this.props.children` but can be any list of elements which
			 *	                          might have this type
			 * @return {Array} `props` objects of elements with this component type
			 */
			getAllProps(elements) {
				let component = this;
				return _.map(component.getAll(elements), 'props');
			},

			getAllAsProps(props, elements) {
				let component = this;
				return _.map(_.flatten([_.get(props, definition.propName, [])]), (childProp) => {
					if (_.isPlainObject(childProp)) {
						return _.assign({}, childProp, {children: childProp.children || childProp.text});
					} else {
						return {
							children: childProp
						};
					}
				}).concat(component.getAllProps(elements));
			}
		}
	});
}
