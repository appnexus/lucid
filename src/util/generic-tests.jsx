import sinon from 'sinon';
import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import describeWithDOM  from './describe-with-dom';
import _ from 'lodash';
import * as lucid from '../index';

// Common tests for all our components
export function common(Component, {
	getDefaultProps = _.noop,
	selector = null,
	exemptFunctionProps = [],
} = {}) {

	function generateDefaultProps(props={}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			let style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(<Component {...generateDefaultProps()} style={style}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const rootStyle = rootWrapper.prop('style');
			assert(_.every(style, (val, key) => val === rootStyle[key]), 'root style must contain passed styles');
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(<Component {...generateDefaultProps()} className={expectedClass}/>);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${expectedClass}'`);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${Component.displayName}'`);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			const rootWrapper = selector ? wrapper.find(selector).first() : wrapper.first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '))
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, className => {
				assert(_.includes(className, `lucid-${Component.displayName}`), `${className} must be scoped`);
			});
		});

		it('should only use onX convention for function proptypes', () => {

			_.forEach(Component.propTypes, (value, key) => {
				// See the following tests from the React source code to figure out how
				// this works: https://github.com/facebook/react/blob/v0.14.7/src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js
				let props = {};
				props[key] = _.noop;

				// Here we test the validation against a noop function, if it doesn't
				// have an error, then it's probably a function. The other possible is
				// that it's an `any`, so we'll test for that below
				const isProbablyFunction = !(value(props, key) instanceof Error);

				// We'll also test out a string to see if that valid
				props[key] = '';

				const isAny = isProbablyFunction && !(value(props, key) instanceof Error);

				// If it's probably a function, and it's not `any`, then we make sure
				// it starts with `on`

				if (isProbablyFunction) {
					assert(isAny || _.startsWith(key, 'on') || _.includes(exemptFunctionProps, key), `${key} must follow onX convention`);
				}

			});

		});

		// Only run this test if it's a public component
		if (!Component._lucidIsPrivate) {
			it('should be available as an exported module from index.js', () => {
				assert(lucid[Component.displayName]);
			});
		}
	});
}

// Common tests for all our icon components
export function icons(Component) {
	describeWithDOM('[icon]', () => {
		it('should pass through isBadge prop to underlying Icon component', () => {
			const wrapper = mount(<Component isBadge={true} />);
			const classNames = wrapper.find('svg').prop('className').split(' ');
			const targetClassName = 'lucid-Icon-is-badge';
			assert(_.includes(classNames, targetClassName), `'${classNames}' should include '${targetClassName}'`);
		});
	});
}

// Common tests for all control components
export function controls(Component, { callbackName, controlSelector , eventType }) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describeWithDOM('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialProp: expectedSpecialProp,
				[callbackName]: sinon.spy(),
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const { props: { specialProp }, event } = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(specialProp, expectedSpecialProp, 'incorrect or missing specialProp');
		});
	});
}
