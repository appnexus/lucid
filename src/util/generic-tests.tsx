import sinon from 'sinon';
import { parse } from 'path';
import React from 'react';
import PropTypes from 'prop-types';
import { mount, shallow, render } from 'enzyme';
import assert from 'assert';
import _, { each, omit, keys, includes } from 'lodash';
import glob from 'glob';
import { addons, mockChannel } from '@storybook/addons';
import timekeeper from 'timekeeper';

addons.setChannel(mockChannel());

interface ICommonConfig {
	getDefaultProps?: () => {};
	exemptFunctionProps?: string[];
	exemptChildComponents?: string[];
	selectRoot?: any;
	noExport?: boolean;
}

// Common tests for all our components
export function common(Component: any, config: ICommonConfig = {} as any) {
	const {
		getDefaultProps = _.constant({}),
		exemptFunctionProps = [],
		exemptChildComponents = [],
		selectRoot = _.identity,
		noExport = false,
	} = config;

	function generateDefaultProps(props = {}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe('[common]', () => {
		if (!Component) {
			throw new Error('An undefined component was passed to generic tests.');
		}

		if (Component._isLucidHybridComponent) {
			throw new Error(
				`You're trying to run generic tests on a hybrid component which is bad and won't work and will make you cry. Check your spec files for ${Component.displayName} and import the raw component instead of the hybrid version.`
			);
		}

		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			const style = {
				backgroundColor: '#f0f',
			};
			const wrapper = shallow(
				<Component {...generateDefaultProps()} style={style} />,
				{ disableLifecycleMethods: true }
			);

			const rootWrapper = selectRoot(wrapper).first();
			const rootStyle = rootWrapper.prop('style');
			assert(
				_.every(style, (val, key) => val === rootStyle[key]),
				'root style must contain passed styles'
			);
		});

		it('should pass through `className`', () => {
			const expectedClass = 'rAnDoM';
			const wrapper = shallow(
				<Component {...generateDefaultProps()} className={expectedClass} />,
				{ disableLifecycleMethods: true }
			);
			const rootWrapper = selectRoot(wrapper).first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(
				_.includes(classNames, expectedClass),
				`'${classNames}' should include '${expectedClass}'`
			);
		});

		it('should have an application scoped base class', () => {
			const expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />, {
				disableLifecycleMethods: true,
			});
			const rootWrapper = selectRoot(wrapper).first();
			const classNames = rootWrapper.prop('className').split(' ');

			assert(
				_.includes(classNames, expectedClass),
				`'${classNames}' should include '${Component.displayName}'`
			);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />, {
				disableLifecycleMethods: true,
			});
			const rootWrapper = selectRoot(wrapper).first();
			const parentClasses = rootWrapper.prop('className').split(' ');
			const childrenClasses = rootWrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '));
			}, []);

			const allClasses = parentClasses.concat(childrenClasses);

			_.forEach(allClasses, (className) => {
				assert(
					_.includes(className, `lucid-${Component.displayName}`),
					`${className} must be scoped`
				);
			});
		});

		describe('function propTypes', () => {
			const funcProps = _.pickBy(
				Component.propTypes,
				(propType) => propType === PropTypes.func
			);

			_.forEach(funcProps, (propType, propName) => {
				it(`${propName} should only use onX convention for function proptypes`, () => {
					assert(
						_.startsWith(propName, 'on') ||
							_.includes(exemptFunctionProps, propName),
						`${propName} must follow onX convention`
					);
				});
			});
		});

		describe('child components', () => {
			// Child components are all function types which start with a capital letter
			const childComponents = _.pickBy(Component, (value, key) => {
				return /^[A-Z]/.test(key) && _.isFunction(value);
			});

			describe('propNames in propTypes', () => {
				_.flow(
					(x) => _.map(x, 'propName'),
					(x) => _.compact(x),
					(x) => _.flatMap(x, _.castArray),
					(x) =>
						_.reject(x, (propName) =>
							_.includes(exemptChildComponents, propName)
						),
					(x) =>
						_.forEach(x, (propName) => {
							it(`should include ${propName} in propTypes`, () => {
								assert(
									Component.propTypes[propName],
									`must include ${propName} in propTypes`
								);
							});
						})
				)(childComponents);
			});
		});

		describe('example testing', () => {
			const fileNames = glob.sync(
				`./src/components/**/${Component.displayName}/*.stories.@(j|t)sx`
			);
			each(fileNames, (path) => {
				const lib = require('../../' + path.replace('.tsx', ''));

				each(omit(lib, ['default']), (Story, name) => {
					it(`should match snapshot(s) for ${name}`, () => {
						let result;
						try {
							result = render(<Story {...Story.args} />, {
								disableLifecycleMethods: true,
							});
							expect(result).toMatchSnapshot();
						} catch (err) {
							expect(err).toMatchSnapshot();
						}
					});
				});
			});

			// Support for older examples
			const exampleFileNames = glob.sync(
				`./src/components/**/${Component.displayName}/examples/*.@(j|t)sx`
			);
			_.each(exampleFileNames, (path) => {
				const lib = require('../../' + path.replace('.tsx', ''));

				const Example = lib.default;

				const title = parse(path).name;
				it(`should match snapshot(s) for ${title}`, () => {
					const shallowExample = shallow(<Example />, {
						disableLifecycleMethods: true,
					});

					// If the root of the example is an instance of the Component under test, snapshot it.
					// Otherwise, look under the root for instances of the Component and snapshot those.
					if (shallowExample.is(Component.displayName)) {
						expect(
							shallow(<Component {...shallowExample.props()} />, {
								disableLifecycleMethods: true,
							})
						).toMatchSnapshot();
					} else {
						shallowExample.find(Component.displayName).forEach((example) => {
							expect(
								shallow(<Component {...example.props()} />, {
									disableLifecycleMethods: true,
								})
							).toMatchSnapshot();
						});
					}
				});
			});
		});
	});
}

// Common tests for all our icon components

interface IIconConfig {
	includeInitialState: boolean;
}

export function icons(Component: any, config: IIconConfig = {} as any) {
	// The default expectation is for every Icon to omit `initailState`,
	// if it is passed through to the underlying element
	const { includeInitialState = false } = config;

	describe('[icon]', () => {
		it('should almost always omit the `initialState` key', () => {
			const wrapper = shallow(<Component initialState={{ testState: true }} />);
			const rootProps = keys(wrapper.first().props());

			expect(includes(rootProps, 'initialState')).toBe(includeInitialState);
		});

		it('should add the correct class for isClickable', () => {
			const wrapper = mount(<Component isClickable={true} />);
			const targetClassName = 'lucid-Icon-is-clickable';
			assert(
				wrapper.find('svg').hasClass(targetClassName),
				`Missing '${targetClassName}' class`
			);
		});
	});
}

// Common tests for all control components
export function controls(
	Component: any,
	{ callbackName, controlSelector, eventType, additionalProps = {} }: any
) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describe('[control]', () => {
		it('should callback with `event` and `props`', () => {
			const expectedSpecialProp = 32;
			const props = {
				specialprop: expectedSpecialProp,
				[callbackName]: sinon.spy(),
				...additionalProps,
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).first().simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const {
				props: { specialprop },
				event,
			}: any = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(
				specialprop,
				expectedSpecialProp,
				'incorrect or missing specialProp'
			);
		});
	});
}

// Common tests for all Functional Components
//
// These tests are intended to help us make sure our FCs are shaped corrected.
// They are necessary because there isn't a perfect way to get the defaultProps
// to be factored in correctly yet with React/TypeScript:
// https://github.com/DefinitelyTyped/DefinitelyTyped/issues/30695#issuecomment-474780159
export function functionalComponents(FC: any) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describe('[functionalComponent]', () => {
		it('should have the correct `peek` properties', () => {
			expect(FC.propName === undefined || typeof FC.propName === 'string').toBe(
				true
			);
			expect(
				FC._isPrivate === undefined || typeof FC._isPrivate === 'boolean'
			).toBe(true);

			expect(typeof FC.peek).toBe('object');
			expect(typeof FC.peek.description).toBe('string');
			expect(
				FC.peek.extend === undefined || typeof FC.peek.extend === 'string'
			).toBe(true);
			expect(
				FC.peek.extend === undefined || typeof FC.peek.extend === 'string'
			).toBe(true);
			expect(
				FC.peek.categories === undefined || Array.isArray(FC.peek.categories)
			).toBe(true);
			expect(
				FC.peek.madeFrom === undefined || Array.isArray(FC.peek.madeFrom)
			).toBe(true);
		});
	});
}

export const mockDate = (dateString: string) => {
	timekeeper.freeze(new Date(dateString));
};
