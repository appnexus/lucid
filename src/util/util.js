import _ from 'lodash';
import React from 'react';

export const isDevMode = (function isReactInDev() {
	try {
		React.createClass({});
	} catch(e) {
		if (e.message.indexOf('render') !== 0) {
			return true;  // A nice, specific error message
		}
	}

	return false;  // should never happen, but play it safe.
})();

export const isNode = typeof process === 'object' && process.title === 'node';

export const logger = (function() {
	return checkIsDev() ? {
		log,
		warn,
		error
	} : {
		log: _.noop,
		warn: _.noop,
		error: _.noop
	};
})();

export function checkIsDev() {
	return isDevMode &&
		typeof window !== 'undefined' &&
		!isNode &&
		typeof console !== 'undefined';
}

function log(...args) {

	console.log(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function warn(...args) {

	console.warn(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}

function error(...args) {

	console.error(...args);

	try {
		// --- Welcome to debugging Lucid ---
		// This error was thrown as a convenience so that you can use this
		// stack to find the callsite that caused this warning to fire.
		throw new Error(args[0]);
	} catch (x) {}

}
