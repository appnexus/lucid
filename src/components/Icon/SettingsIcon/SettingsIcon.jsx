import _ from 'lodash';
import React from 'react';
import Icon from '../Icon';
import { lucidClassNames } from '../../../util/style-helpers';
import { createClass as createReactClass } from '../../../util/component-types';

const cx = lucidClassNames.bind('&-SettingsIcon');

/**
 *
 * {"categories": ["visual design", "icons"], "extend": "Icon", "madeFrom": ["Icon"]}
 *
 * A settings icon.
 */
const SettingsIcon = createReactClass({
	displayName: 'SettingsIcon',
	propTypes: {
		...Icon.propTypes,
	},

	render() {
		const { className, ...passThroughs } = this.props;

		return (
			<Icon
				{...passThroughs}
				{..._.pick(passThroughs, _.keys(Icon.propTypes))}
				className={cx('&', className)}
			>
				<path d="M14.146,6.836l-1.939-0.356c-0.188-0.035-0.254-0.192-0.146-0.351l1.111-1.647c0.107-0.158,0.084-0.399-0.051-0.534 l-1.066-1.066C11.92,2.746,11.678,2.723,11.52,2.83L9.873,3.941c-0.158,0.106-0.316,0.04-0.352-0.147L9.166,1.855 C9.131,1.668,8.945,1.514,8.754,1.514H7.246c-0.191,0-0.375,0.154-0.41,0.342L6.479,3.794c-0.033,0.188-0.191,0.254-0.35,0.147 L4.48,2.83C4.322,2.723,4.082,2.746,3.947,2.881L2.881,3.947C2.746,4.083,2.723,4.323,2.828,4.481l1.113,1.647 c0.105,0.158,0.041,0.315-0.148,0.351L1.855,6.836C1.668,6.871,1.514,7.057,1.514,7.247v1.508c0,0.191,0.154,0.376,0.342,0.41 l1.939,0.357c0.188,0.034,0.254,0.192,0.146,0.35l-1.113,1.648c-0.105,0.159-0.082,0.398,0.053,0.533l1.066,1.066 c0.135,0.136,0.375,0.158,0.533,0.053l1.648-1.112c0.158-0.106,0.316-0.041,0.35,0.147l0.357,1.938 c0.035,0.188,0.219,0.342,0.41,0.342h1.508c0.191,0,0.377-0.153,0.412-0.342l0.355-1.938c0.035-0.188,0.193-0.255,0.352-0.147 l1.646,1.112c0.158,0.105,0.4,0.083,0.535-0.053l1.066-1.064c0.135-0.137,0.158-0.376,0.051-0.535l-1.111-1.648 c-0.107-0.157-0.041-0.315,0.146-0.35l1.939-0.357c0.188-0.034,0.34-0.219,0.34-0.41V7.247C14.486,7.057,14.334,6.871,14.146,6.836 z M8,11.086c-1.701,0-3.082-1.378-3.082-3.08c0-1.701,1.381-3.081,3.082-3.081s3.082,1.379,3.082,3.081 C11.082,9.708,9.701,11.086,8,11.086z" />
			</Icon>
		);
	},
});

export default SettingsIcon;
