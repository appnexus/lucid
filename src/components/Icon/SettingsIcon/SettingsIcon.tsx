import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import Icon, { IIconProps } from '../Icon';
import { lucidClassNames } from '../../../util/style-helpers';

const cx = lucidClassNames.bind('&-SettingsIcon');

export const iconPropTypes = {
	/** Size variations of the icons. \`size\` directly effects height and width
		but the developer should also be conscious of the relationship with
		\`viewBox\`. */
	size: PropTypes.number,

	/** Size handles width and height, whereas \`width\` can manually override the width that would be set by size. */
	width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

	/** Size handles width and height, whereas \`height\` can manually override the height that would be set by size. */
	height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),

	/** \`viewBox\` is very important for SVGs. You can think of \`viewBox\` as
		the "artboard" for our SVG while \`size\` is the presented height and
		width. */
	viewBox: PropTypes.string,

	/** Sets the color of the Icon.  May not be applicable for icons that are tied
		to specific colors (e.g. DangerIcon). */
	color: PropTypes.oneOf([
		'neutral-dark',
		'neutral-light',
		'primary',
		'white',
		'success',
		'warning',
		'secondary-one',
		'secondary-two',
		'secondary-three',
	]),

	/** Any valid SVG aspect ratio. */
	aspectRatio: PropTypes.string,

	/** Adds styling that makes the icon appear clickable. */
	isClickable: PropTypes.bool,

	/** Adds styling that makes the icon appear disabled.  Also forces
		isClickable to be false. */
	isDisabled: PropTypes.bool,

	/** Called when the user clicks the \`Icon\`. Signature:
		\`({event, props}) => {}\` */
	onClick: PropTypes.func,

	/** Called when the user clicks an active, clickable \`Icon\`. Signature:
		\`({event, props}) => {}\` */
	onSelect: PropTypes.func,

	/** Any valid React children. */
	children: PropTypes.element,

	/** Classes that are appended to the component defaults. This prop is run
		through the \`classnames\` library. */
	className: PropTypes.string,
};

export const SettingsIcon = ({ className, ...passThroughs }: IIconProps) => {
	return (
		<Icon
			{..._.omit(passThroughs, ['initialState'])}
			className={cx('&', className)}
		>
			<path d='M2.254 13.052l1.733-.999A5.702 5.702 0 0 0 6.5 13.501V15.5h3v-1.999a5.708 5.708 0 0 0 2.513-1.451l1.733.999 1.5-2.599-1.733-.998c.25-.951.25-1.951 0-2.902l1.732-1.001-1.499-2.599-1.733 1A5.707 5.707 0 0 0 9.5 2.502V.5h-3v2.002a5.696 5.696 0 0 0-2.513 1.45l-1.726-1L.76 5.553l1.727.997a5.708 5.708 0 0 0 0 2.902L.755 10.453l1.499 2.599z' />
			<circle cx='8' cy='8' r='2.5' />
		</Icon>
	);
};

SettingsIcon.displayName = 'SettingsIcon';

SettingsIcon.propTypes = iconPropTypes;

SettingsIcon.defaultProps = Icon.defaultProps;

export default SettingsIcon;
