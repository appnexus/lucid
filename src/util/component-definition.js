import React from 'react';
import _ from 'lodash';
import { createChildComponent } from './child-component';

export function createLucidComponentDefinition(componentDefinition) {
	let {
		arrayOf,
		shape,
		oneOfType,
		node
	} = React.PropTypes;
	let reducers = _.get(componentDefinition, 'reducers');
	let childProps = _.get(componentDefinition, 'childProps', {});
	let childPropsList = _.reduce(childProps, (list, value, key) => list.concat({
		propTypes: value,
		baseName: key,
		propName: key,
		displayName: _.flatten([_.get(componentDefinition, 'displayName', [])]).concat([key]).join('.'),
	}), []);

	return _.assign({}, _.omit(componentDefinition, ['reducers', 'childProps']), {
		propTypes: _.assign(
			{},
			_.get(componentDefinition, 'propTypes'),
			_.reduce(
				childPropsList,
				(newPropTypes, childProp) => _.assign(
					{},
					newPropTypes,
					{[childProp.propName]: oneOfType([arrayOf(shape(childProp.propTypes)), shape(childProp.propTypes), node])}
				),
				{}
			)
		),
		statics: _.assign(
			{},
			_.get(componentDefinition, 'statics'),
			(reducers ? {reducers} : null),
			_.reduce(
				childPropsList,
				(exportedChildComponents, childProp) => _.assign(
					{},
					exportedChildComponents,
					{[childProp.baseName]: createChildComponent(childProp)}
				),
				{}
			)
		)
	});
}
