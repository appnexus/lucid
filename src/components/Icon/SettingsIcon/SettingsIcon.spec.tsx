import { icons, common } from '../../../util/generic-tests';

import SettingsIcon from './SettingsIcon';

describe('SettingsIcon', () => {
	common(SettingsIcon);
	icons(SettingsIcon);
});
