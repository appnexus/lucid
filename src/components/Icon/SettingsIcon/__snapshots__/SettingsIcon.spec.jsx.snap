// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsIcon [common] example testing should match snapshot(s) for 1.default 1`] = `
<Icon
  aspectRatio="xMidYMid meet"
  className="lucid-SettingsIcon"
  isBadge={false}
  isClickable={false}
  isDisabled={false}
  size={16}
  viewBox="0 0 16 16"
>
  <path
    d="M14.146 6.836l-1.94-.356c-.187-.035-.253-.192-.145-.35l1.112-1.648c.107-.158.084-.4-.05-.534l-1.067-1.066c-.135-.136-.377-.16-.535-.052L9.873 3.94c-.158.107-.316.04-.352-.146l-.354-1.94c-.035-.186-.22-.34-.412-.34H7.246c-.19 0-.375.154-.41.342L6.48 3.794c-.034.188-.192.254-.35.147L4.48 2.83c-.158-.107-.398-.084-.533.05L2.88 3.948c-.134.136-.157.376-.052.534L3.94 6.13c.106.158.042.315-.147.35l-1.938.358c-.187.035-.34.22-.34.41v1.51c0 .19.153.375.34.41l1.94.356c.188.034.254.192.146.35L2.83 11.52c-.105.16-.082.398.053.533l1.067 1.066c.135.135.375.157.533.052l1.648-1.112c.158-.106.316-.04.35.147l.357 1.938c.035.188.22.342.41.342h1.508c.19 0 .377-.153.412-.342l.355-1.938c.035-.188.193-.255.352-.147l1.646 1.112c.158.105.4.083.535-.053l1.066-1.065c.134-.137.157-.376.05-.535l-1.11-1.648c-.108-.157-.042-.315.145-.35l1.94-.357c.187-.034.34-.22.34-.41V7.247c0-.19-.15-.376-.34-.41zM8 11.086c-1.7 0-3.082-1.378-3.082-3.08 0-1.7 1.38-3.08 3.082-3.08s3.082 1.378 3.082 3.08c0 1.702-1.38 3.08-3.082 3.08z"
  />
</Icon>
`;

exports[`SettingsIcon [common] example testing should match snapshot(s) for 2.clickable 1`] = `
<Icon
  aspectRatio="xMidYMid meet"
  className="lucid-SettingsIcon"
  isBadge={false}
  isClickable={true}
  isDisabled={false}
  size={16}
  viewBox="0 0 16 16"
>
  <path
    d="M14.146 6.836l-1.94-.356c-.187-.035-.253-.192-.145-.35l1.112-1.648c.107-.158.084-.4-.05-.534l-1.067-1.066c-.135-.136-.377-.16-.535-.052L9.873 3.94c-.158.107-.316.04-.352-.146l-.354-1.94c-.035-.186-.22-.34-.412-.34H7.246c-.19 0-.375.154-.41.342L6.48 3.794c-.034.188-.192.254-.35.147L4.48 2.83c-.158-.107-.398-.084-.533.05L2.88 3.948c-.134.136-.157.376-.052.534L3.94 6.13c.106.158.042.315-.147.35l-1.938.358c-.187.035-.34.22-.34.41v1.51c0 .19.153.375.34.41l1.94.356c.188.034.254.192.146.35L2.83 11.52c-.105.16-.082.398.053.533l1.067 1.066c.135.135.375.157.533.052l1.648-1.112c.158-.106.316-.04.35.147l.357 1.938c.035.188.22.342.41.342h1.508c.19 0 .377-.153.412-.342l.355-1.938c.035-.188.193-.255.352-.147l1.646 1.112c.158.105.4.083.535-.053l1.066-1.065c.134-.137.157-.376.05-.535l-1.11-1.648c-.108-.157-.042-.315.145-.35l1.94-.357c.187-.034.34-.22.34-.41V7.247c0-.19-.15-.376-.34-.41zM8 11.086c-1.7 0-3.082-1.378-3.082-3.08 0-1.7 1.38-3.08 3.082-3.08s3.082 1.378 3.082 3.08c0 1.702-1.38 3.08-3.082 3.08z"
  />
</Icon>
`;

exports[`SettingsIcon [common] example testing should match snapshot(s) for 3.disabled 1`] = `
<Icon
  aspectRatio="xMidYMid meet"
  className="lucid-SettingsIcon"
  isBadge={false}
  isClickable={false}
  isDisabled={true}
  size={16}
  viewBox="0 0 16 16"
>
  <path
    d="M14.146 6.836l-1.94-.356c-.187-.035-.253-.192-.145-.35l1.112-1.648c.107-.158.084-.4-.05-.534l-1.067-1.066c-.135-.136-.377-.16-.535-.052L9.873 3.94c-.158.107-.316.04-.352-.146l-.354-1.94c-.035-.186-.22-.34-.412-.34H7.246c-.19 0-.375.154-.41.342L6.48 3.794c-.034.188-.192.254-.35.147L4.48 2.83c-.158-.107-.398-.084-.533.05L2.88 3.948c-.134.136-.157.376-.052.534L3.94 6.13c.106.158.042.315-.147.35l-1.938.358c-.187.035-.34.22-.34.41v1.51c0 .19.153.375.34.41l1.94.356c.188.034.254.192.146.35L2.83 11.52c-.105.16-.082.398.053.533l1.067 1.066c.135.135.375.157.533.052l1.648-1.112c.158-.106.316-.04.35.147l.357 1.938c.035.188.22.342.41.342h1.508c.19 0 .377-.153.412-.342l.355-1.938c.035-.188.193-.255.352-.147l1.646 1.112c.158.105.4.083.535-.053l1.066-1.065c.134-.137.157-.376.05-.535l-1.11-1.648c-.108-.157-.042-.315.145-.35l1.94-.357c.187-.034.34-.22.34-.41V7.247c0-.19-.15-.376-.34-.41zM8 11.086c-1.7 0-3.082-1.378-3.082-3.08 0-1.7 1.38-3.08 3.082-3.08s3.082 1.378 3.082 3.08c0 1.702-1.38 3.08-3.082 3.08z"
  />
</Icon>
`;
